CORE-10335 ledger migration module

MrMatthewLayton opened the pull request:
The current changes add conversion of ledger-common and ledger-consensual to Java

suhas-srivastava committed:
CORE-10335: Transferred TransactionMetadata.kt to java directory.

suhas-srivastava committed:
CORE-10335: Transferred TransactionMetadata.kt to java directory.

suhas-srivastava committed:
CORE-10335: Migrated code for TransactionMetadata.kt to java.

suhas-srivastava committed:
CORE-10335: Migrated code for TransactionMetadata.kt to java.

suhas-srivastava committed:
CORE-10335: Migrated TransactionSignatureService.kt to java directory.

suhas-srivastava committed:
CORE-10335: Migrated TransactionSignatureService.kt to java directory.

suhas-srivastava committed:
CORE-10335: Migrated code of TransactionSignatureService.kt to java.

suhas-srivastava committed:
CORE-10335: Migrated code of TransactionSignatureService.kt to java.

MrMatthewLayton committed:
Converted ledger-common and ledger-consensual to Java.

MrMatthewLayton committed:
Converted ledger-common and ledger-consensual to Java.

MrMatthewLayton committed:
Added TODO comment for review.

MrMatthewLayton committed:
Added TODO comment for review.

chrisr3 commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/PrivacySalt.java:
```diff
@@ -0,0 +1,23 @@
+ * A privacy salt is required to compute  a nonce per transaction component in order to ensure that an adversary cannot
+ * use brute force techniques and reveal the content of a Merkle-leaf hashed value.
+ */
+@CordaSerializable
```
Why is this `@CordaSeriaizable` now? Wasn't the implementation _already_ `@CordaSerializable`, and isn't that enough?

MrMatthewLayton replied:
@chrisr3 **PrivacySalt.kt** was annotated `@CordaSerializable`, so I did the same here. Shouldn't we annotate on the interface (where applicable) as a general rule? i.e. if it's on the interface, it will apply to any/all implementations.

lankydan replied:
Whether we keep the annotation on interfaces is TBD atm.

MrMatthewLayton committed:
[WIP] Migrating ledger-utxo to Java.

MrMatthewLayton committed:
[WIP] Migrating ledger-utxo to Java.

MrMatthewLayton committed:
Migrating all ledger-utxo to Java.

MrMatthewLayton committed:
Migrating all ledger-utxo to Java.

blsemo commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionMetadata.java:
```diff
@@ -1,71 +1,75 @@
-    /**
-     * Gets the version of the platform at the time of the creation of the transaction.
-     *
-     * @return The platform version.
```
What happened to the platform version?

vlajos replied:
Probably the base branch was cut before platform version got added to the main branch.
Anyway, it is needed.

blsemo commented on ledger/ledger-common/src/test/java/net/corda/v5/ledger/common/transaction/TransactionVerificationExceptionJavaApiTest.java:
```diff
@@ -4,41 +4,44 @@
 import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-public class TransactionSignatureExceptionJavaApiTest {
```
Erm - I thought Java API tests can just be deleted once the API is Java? They're only purpose was to keep the API usable from Java.

MrMatthewLayton replied:
Isn't there still value in having the tests? The tests not only prove that the code is callable from Java, but they also prove that they behave as expected. That said, if we're testing the real, un-mocked logic in runtime, then these aren't strictly necessary.

lankydan replied:
There is no benefit in these tests, everything should be tested in runtime-os. Even the mocks in these tests are overkill for the original usage of testing calling kotlin from java...

@chrisr3 has kept them in his migration and I imagine others are probably doing the same.

I guess the only argument I can make for them is that it makes it clear if we break an existing API. However, no one has really been adding Java API tests along with new APIs in the last year.

blsemo replied:
They really should be deleted. There is literally *no point* in testing that mocks behave as they've been set up to behave.

blsemo commented on ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/transaction/ConsensualSignedTransaction.java:
```diff
@@ -0,0 +1,58 @@
+ * Adding or removing a signature does not change the transaction ID.
+ */
+@DoNotImplement
+public interface ConsensualSignedTransaction {
```
What version of the API have you been porting this off? This does not seem to be in line with the latest API, it's missing the `TransactionWithMetadata` base interface.

MrMatthewLayton replied:
I've been pulling and rebasing from `release/os/5.0` daily.

blsemo reviewed (commented):
@MrMatthewLayton Can you please check this is a port of the latest version of the ledger API? This looks at least a week or two out of date.

vlajos commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionWithMetadata.java:
```diff
@@ -0,0 +1,29 @@
+ * Defines metadata properties of transactions common across difference ledger implementations.
+ */
+@DoNotImplement
+// TODO : Should this be marked @CordaSerializable? Instances of this will be send over the wire.
```
I think it only gets json serialized.

lankydan replied:
Annotations on interfaces is being discussed atm.

lankydan replied:
Please remove the TODOs around `@CordaSerializable` for now.

vlajos commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionVerificationException.java:
```diff
@@ -0,0 +1,36 @@
+    public SecureHash getTransactionId() {
+        return this.transactionId;
+    }
+}
```
This class got deleted/renamed in the meantime.
(And the new one is java.)

vlajos replied:
(The new one is the TransactionSignatureException)

lankydan replied:
Still needs updating.

vlajos commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoLedgerTransaction.java:
```diff
@@ -0,0 +1,296 @@
+     *
+     * @return Returns the reference input state refs associated with the current {@link UtxoLedgerTransaction}.
+     */
+    // TODO : Rename to getReferenceInputStateRefs
```
Regarding adding back the `input` to the name of these functions:
We had a short discussion on slack in December where it seemed that the shorter (without `input`) is the preferred option.

lankydan replied:
Did we have a meeting about this? I thought we agreed on `getReferenceStateRefs` / `getReferenceStates`. Either way is the TODO resolved and can be removed?

MrMatthewLayton committed:
Few discrepancies found while porting runtime.

MrMatthewLayton committed:
Few discrepancies found while porting runtime.

Commit:
Bump shared lib version (#885)

relyafi committed:
CORE-7292 Update notary client plugin KDocs (#886)

chrisr3 committed:
CORE-10333: Remove Kotlin "utility" functions from application module. (#877)

yift-r3 committed:
CORE-10124: Make the request state and event queue (#865)

* CORE-10124: Make the request state and event queue

* Apply review comments

nargas-ritu committed:
NOTICK: Snyk waivers for Gecko 2.0

Commit:
Update .snyk

nargas-ritu committed:
Merge pull request #892 from corda/snyk_waivers-Gecko2.0

NOTICK: Snyk waivers for Gecko 2.0

lankydan commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/NotaryLookup.java:
```diff
@@ -0,0 +1,47 @@
+    @NotNull
+    @Suspendable
+    // TODO : Consider List<NotaryInfo>
+    Collection<NotaryInfo> getNotaryServices();
```
I'd say `List` or `Set`. @relyafi can you make a call on this.

relyafi replied:
If its suspendable, doesn't that imply it needs to be serializable, and IIRC aren't there issues with sets, maps etc. with serializing? Or am I getting confused with something else? 

Is there a reason not to leave it as a collection, given it's an interface? In practice I suspect it doesn't matter too much, given there's not likely to be too many notary services on a network anyway.

MrMatthewLayton replied:
@lankydan @relyafi I'll leave the TODO there for the time being. There are a few in the API for consideration, though we probably want to address them sooner rather than later if they're API concerns.

lankydan replied:
If we're going to leave the TODO, we should raise a ticket for it and put it highest so it can get scoped. We either do this now or we don't. If we put it as a "should" it won't get done regardless.

relyafi replied:
Fair point.

So going back to my original comment:

_Is there a reason not to leave it as a collection, given it's an interface?_

Can we start with answering that?

lankydan commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/Party.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * Represents a well-known party identity.
+ */
+// TODO : Rename owningKey to signingKey
```
@blsemo + @mnesbit can be make a call on this TODO

blsemo replied:
Renaming to signing key makes sense - given that we use Party only to identify notaries (and possbibly issuers) which don't own any states.

MrMatthewLayton replied:
@lankydan I spoke with @mnesbit. He said for review in a future ticket, but not now. Bigger question is are we keeping `Party` anyway?

lankydan replied:
Raise a Ticket and remove the TODO. Don't know about `Party`.

lankydan commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/Party.java:
```diff
@@ -0,0 +1,103 @@
+     * @return Returns true if the specified object is equal to the current object; otherwise, false.
+     */
+    @Override
+    public boolean equals(@Nullable final Object obj) {
```
Why do we need 2 `equals` methods? Are the Intellij generated ones not good enough here?

MrMatthewLayton replied:
@lankydan The IntelliJ ones will probably suffice, so I'll refactor.

lankydan commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionNoAvailableKeysException.java:
```diff
@@ -0,0 +1,23 @@
+     * @param message The details of the current exception to throw.
+     * @param cause The underlying cause of the current exception.
+     */
+    public TransactionNoAvailableKeysException(
+            @NotNull final String message,
+            @Nullable final Throwable cause) {
+        super(message, cause);
```
Formatting here is a bit off.

MrMatthewLayton replied:
@lankydan The formatting is IntelliJ's default (as per the style/formatting convention discussion on Thursday). It is to distinguish between the constructor/method parameters, and the body. Hitting <kbd>command</kbd> + <kbd>option</kbd> + <kbd>l</kbd> and then <kbd>option</kbd> + <kbd>control</kbd> + <kbd>o</kbd> will auto-format and organise imports.

lankydan replied:
The standard intellij formatting for Java is awful. If you look through the Java code the last `)` is all over the place in our APIs.

lankydan commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionSignatureService.java:
```diff
@@ -1,57 +1,68 @@
-    fun verifySignature(transaction: TransactionWithMetadata, signatureWithMetadata: DigitalSignatureAndMetadata)
-}
\ No newline at end of file
+    void verifySignature(
```
Should really have a `@throws` of what exceptions it throws. a `RuntimeException` at a minimum will suffice.

lankydan commented on ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionVerificationException.java:
```diff
@@ -0,0 +1,36 @@
+    @NotNull
+    private final SecureHash transactionId;
+
+    public TransactionVerificationException(
+            @NotNull final SecureHash transactionId,
+            @NotNull final String message,
+            @Nullable final Throwable cause) {
+        super(message, cause);
+        this.transactionId = transactionId;
```
Formatting on end `)`

MrMatthewLayton replied:
@lankydan Not sure what you mean?

lankydan commented on ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/ConsensualState.java:
```diff
@@ -0,0 +1,49 @@
+     *
+     * @param transaction The transaction in which the current consensual state will be created.
+     */
+    void verify(@NotNull final ConsensualLedgerTransaction transaction);
```
Remove this `final`.

lankydan commented on ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/transaction/ConsensualTransactionBuilder.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    @NotNull
+    // TODO : Consider alignment with UTXO addState and addStates.
+    ConsensualTransactionBuilder withStates(ConsensualState... states);
```
I agree in making this `add` instead of `with`. @blsemo what do you think?

If we do this, please create a separate ticket and quickly do it after this task.

blsemo replied:
Yes, that makes sense - both points.

lankydan commented on ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/transaction/ConsensualTransactionValidator.java:
```diff
@@ -0,0 +1,42 @@
+     * @throws Throwable if the {@link ConsensualLedgerTransaction} fails validation.
+     */
+    @Suspendable
+    void checkTransaction(@NotNull ConsensualLedgerTransaction transaction) throws Throwable;
```
Remove the `throws` since it doesn't have to be there and the `@throws` is enough.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/Attachment.java:
```diff
@@ -0,0 +1,82 @@
+ */
+@DoNotImplement
+@CordaSerializable
+public interface Attachment {
```
Honestly, @blsemo @mnesbit we should just delete the `Attachment` interface since we haven't done any implementation towards it. We can re-add it later.

blsemo replied:
I agree in principle - we might even want attachments to be slightly different in C5 compared to C4, so committing to an API without even trying to implement it is dangerous. 
Can you create a ticket for it?

MrMatthewLayton replied:
@lankydan @blsemo @mnesbit If we delete attachment, we will also remove a whole bunch of functionality throughout the API, e.g. `addAttachment` on `UtxoTransactionBuilder`, and `getAttachments`/`getAttachment` on `UtxoLedgerTransaction`, but, I do agree with @lankydan, if we've not really focused on the underlying implementation of how these are going to work, I'd be inclined to remove for now.

To be honest, attachments were always a bit finicky in C4. Perhaps we can come up with something better?

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/Attachment.java:
```diff
@@ -0,0 +1,82 @@
+     * @param output The output stream where the file should be copied to.
+     * @throws FileNotFoundException if the specified path does not exist in the attachment.
+     */
+    void extractFile(@NotNull String path, @NotNull OutputStream output) throws FileNotFoundException;
```
Generally we should be rethrowing all these exceptions so checked exceptions are not used.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/ContractVerificationException.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    public ContractVerificationException(
+            @NotNull final SecureHash transactionId,
+            @NotNull final List<ContractVerificationFailure> failureReasons) {
```
`)` on new line

MrMatthewLayton replied:
@lankydan I think I'm getting what you mean, as per the previous comment too...

Do we have a style preference for this? Personally I think a new line containing `) {` looks terrible, mostly because the Java stylistic convention for `{` is at the end of a line, not on a new line (which in this case could be argued both ways).

```java
public ContractVerificationException(
    @NotNull final SecureHash transactionId,
    @NotNull final List<ContractVerificationFailure> failureReasons
) {
    this.transactionId = transactionId;
    this.failureReasons = failureReasons;
}
```

vs.

```java
public ContractVerificationException(
        @NotNull final SecureHash transactionId,
        @NotNull final List<ContractVerificationFailure> failureReasons) {
    this.transactionId = transactionId;
    this.failureReasons = failureReasons;
}
```

I personally think the latter looks cleaner, but if there's a consensual preference towards the former, I'l change it.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/ContractVerificationException.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    private static String buildExceptionMessage(
+            @NotNull final SecureHash transactionId,
+            @NotNull final List<ContractVerificationFailure> failureReasons) {
```
`)` on new line

MrMatthewLayton replied:
As above.

chrisr3 committed:
CORE-10332: Remove remaining Kotlin from :base and :data. (#889)

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/UtxoLedgerService.java:
```diff
@@ -0,0 +1,128 @@
+    UtxoSignedTransaction finalize(
+            @NotNull UtxoSignedTransaction transaction,
+            @NotNull Iterable<FlowSession> sessions
+    ) throws ContractVerificationException;
```
@mnesbit do we want to keep `throws RuntimeException`s? I just wrote some code and doing so doesn't cause the same issues as declaring throws for a checked exception. So I don't mind, just need something to be consistent to.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserver.java:
````diff
@@ -0,0 +1,86 @@
+ * ```Kotlin
+ * data class ExampleStateK(
+ * override val participants: List<PublicKey>,
+ * val issuer: SecureHash,
````
This code formatting in the javadocs is screwed.

MrMatthewLayton replied:
@lankydan agreed. auto-formatting doesn't seem to like in-line code examples in the docs, and I'm wondering, genuinely how much value there will be in them, as they are sparse throughout the API docs anyway?

lankydan replied:
The inline code examples are extremely helpful imo and we spent a lot of time adding them months ago for Beta 1 or the Cordacon release.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/observer/UtxoToken.java:
```diff
@@ -0,0 +1,123 @@
+    public UtxoToken(
+            @NotNull final UtxoTokenPoolKey poolKey,
+            @NotNull final BigDecimal amount,
+            @Nullable final UtxoTokenFilterFields filterFields) {
```
`)` new line

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/observer/UtxoTokenPoolKey.java:
```diff
@@ -0,0 +1,135 @@
+    public UtxoTokenPoolKey(
+            @Nullable final String tokenType,
+            @NotNull final SecureHash issuerHash,
+            @NotNull final String symbol) {
```
`)` new line

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenClaimCriteria.java:
```diff
@@ -0,0 +1,214 @@
+            @NotNull final SecureHash issuerHash,
+            @NotNull final MemberX500Name notaryX500Name,
+            @NotNull final String symbol,
+            @NotNull final BigDecimal targetAmount) {
```
`)` new line

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenClaimCriteria.java:
```diff
@@ -0,0 +1,214 @@
+    /**
+     * Sets an optional regular expression to match against the {@link ClaimedToken#getTag()} field, or null to match all tags.
+     */
+    public void setTagRegex(@Nullable final String tagRegex) {
```
Remove the `final`s

MrMatthewLayton replied:
@lankydan Whilst there is no value in `final` on interfaces, as they are ignored by the implementation anyway, there is value in having them on implementations (which this is), as this resembles how parameters work in Kotlin. Should these classes ever need to be changed in future, I would hope that they are an indication to the developer that these parameters should not be mutated within the method/constructor body.

@mnesbit @blsemo thoughts?

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenClaimCriteria.java:
```diff
@@ -0,0 +1,214 @@
+     * @param targetAmount   The minimum value for the sum of {@link ClaimedToken#getAmount()} for the selected tokens.
+     */
+    public TokenClaimCriteria(
+            @NotNull final String tokenType,
+            @NotNull final SecureHash issuerHash,
+            @NotNull final MemberX500Name notaryX500Name,
+            @NotNull final String symbol,
+            @NotNull final BigDecimal targetAmount) {
```
Remove the finals

MrMatthewLayton replied:
as above.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/observer/UtxoTokenPoolKey.java:
```diff
@@ -0,0 +1,135 @@
+     * @param symbol The user defined symbol of the tokens in a pool.
+     */
+    public UtxoTokenPoolKey(
+            @Nullable final String tokenType,
```
Remove the finals

MrMatthewLayton replied:
as above.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/observer/UtxoTokenFilterFields.java:
```diff
@@ -0,0 +1,106 @@
+     * @return Returns true if the specified object is equal to the current object; otherwise, false.
+     */
+    @Override
+    public boolean equals(@Nullable final Object obj) {
```
Lots of finals in these methods

MrMatthewLayton replied:
as above.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java:
````diff
@@ -0,0 +1,99 @@
+ * Example of use in Kotlin.
+ * ```Kotlin
+ *
+ * @CordaInject lateinit var tokenSelection: TokenSelection
````
The code in these docs is broken.

MrMatthewLayton replied:
@lankydan See previous comment re. docs.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoLedgerTransaction.java:
```diff
@@ -0,0 +1,288 @@
+     * @param type The type of the {@link ContractState}.
+     * @return Returns all reference input states and state refs that match the specified type.
+     */
+    // TODO : Rename to getReferenceInputStateAndRefs
```
Same with the TODO here and above.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoLedgerTransaction.java:
```diff
@@ -0,0 +1,288 @@
+    // TODO : Rename to getReferenceInputTransactionStates
+    @NotNull
+    default List<TransactionState<?>> getReferenceTransactionStates() {
+        return getReferenceStateAndRefs()
```
Shouldn't affect this PR, but I'd argue we just move most of the default implementations into the actual implementation and streamline the interface code.

MrMatthewLayton replied:
@lankydan Agreed, this should probably be moved to runtime, but do we do that in this PR, or another one?

lankydan replied:
Na leave it for another time.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoLedgerTransaction.java:
```diff
@@ -0,0 +1,288 @@
+     * @param type The type of the {@link ContractState}.
+     * @return Returns all output contract states that match the specified type.
+     */
+    // TODO : Rename to getOutputContractStates
```
What is the resolution of this TODO?

MrMatthewLayton replied:
@blsemo @mnesbit input needed.

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoSignedTransaction.java:
```diff
@@ -0,0 +1,113 @@
+ * Defines a signed UTXO transaction.
+ * <p>
+ * Comparing with {@link UtxoLedgerTransaction}:
+ * - It does not have access to the deserialized details.
+ * - It has direct access to the signatures.
+ * - It does not require a serializer.
```
This doesn't actually format to a list like in Kdocs. You need to use:

```java
 * <ul>
 * <li> It does not have access to the deserialized details.</li>
 * <li> It has direct access to the signatures.</li>
 * <li> It does not require a serializer.</li>
 * </ul>
```

There are probably other instances but this is the first one that I processed through my head.

MrMatthewLayton replied:
Java are the masters of verbosity.

MrMatthewLayton replied:
Actually, with regards to code samples in docs, they are being formatted in markdown. Is that supported in JavaDoc?

lankydan replied:
No, they need to be changed.

Chris did this, which Intellij seems to render nicely.

```java
* Example usage of a {@link SubFlow} not annotated with {@link InitiatingFlow}:
 * <ul>
 * <li>Kotlin:<pre>{@code
 * class MySubFlow(private val existingSession: FlowSession) : SubFlow<String> {
 *
 *     @Suspendable
 *     override fun call(): String {
 *         return existingSession.receive<String>().unwrap { it }
 *     }
 * }
 * }</pre></li>
 * <li>Java:<pre>{@code
 * public class MySubFlow implements SubFlow<String> {
 *
 *     private final FlowSession existingSession;
 *
 *     public MySubFlow(FlowSession existingSession) {
 *         this.existingSession = existingSession;
 *     }
 *
 *     @Suspendable
 *     @Override
 *     public String call() {
 *         return existingSession.receive(String.class).unwrap(it -> it);
 *     }
 * }
 * }</pre></li></ul>
```

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoSignedTransaction.java:
```diff
@@ -0,0 +1,113 @@
+     *
+     * @return Returns the {@link StateRef} instances of the reference inputs for the current transaction.
+     */
+    // TODO : Shouldn't this be called getReferenceInputStateRefs?
```
Is this TODO resolved?

MrMatthewLayton replied:
I've removed them if we're sticking with `getReferenceStateRefs`

lankydan commented on ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/filtered/UtxoFilteredTransaction.java:
```diff
@@ -0,0 +1,118 @@
+     * @return Returns a potentially filtered list of required signatories for the current transaction.
+     */
+    @NotNull
+    UtxoFilteredData<PublicKey> getSignatories();
```
Should we add `@Suspendable` to these methods while we're here? @mnesbit Appreciate that they don't actually suspend.

YashNabar committed:
CORE-10538 Update schemas with registration status reason field (#887)

Updates existing avro and database schemas to add `reason` for registration status.

MrMatthewLayton committed:
Addressed PR feedback

MrMatthewLayton committed:
Merge remote-tracking branch 'origin/CORE-10335-Ledger-migration-module' into CORE-10335-Ledger-migration-module

# Conflicts:
#	gradle.properties
#	ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/Party.java
#	ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionWithMetadata.java
#	ledger/ledger-common/src/test/java/net/corda/v5/ledger/common/transaction/TransactionVerificationExceptionJavaApiTest.java
#	ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/ConsensualState.java
#	ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/transaction/ConsensualSignedTransaction.java
#	ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/transaction/ConsensualTransactionValidator.java
#	ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/Attachment.java
#	ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/StateRef.java
#	ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoLedgerTransaction.java
#	ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoSignedTransaction.java
#	ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/filtered/UtxoFilteredTransaction.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/AbstractMockTestHarness.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/AttachmentJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/ContractJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/ContractStateJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/StateAndRefJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/StateRefJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/TimeWindowJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/TransactionStateJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/UtxoLedgerServiceJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/UtxoLedgerTransactionJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/UtxoSignedTransactionJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/UtxoTransactionBuilderJavaApiTests.java
#	ledger/ledger-utxo/src/test/java/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserverJavaExample.java
#	ledger/ledger-utxo/src/test/kotlin/net/corda/v5/ledger/utxo/observer/UtxoLedgerTokenStateObserverKotlinExample.kt

MrMatthewLayton commented:
THIS BRANCH HAS BEEN SUPERSEDED BY: https://github.com/corda/corda-api/pull/894
